pkg/controller/core/leader_aware_reconciler.go

```diff
@@ -0,0 +1,100 @@
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package core
+
+import (
+	"context"
+	"time"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/utils/ptr"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	config "sigs.k8s.io/kueue/apis/config/v1beta1"
+)
+
+// defaultRequeueDuration defaults the duration used by non-leading replicas
+// to requeue events, so no events are missed over the period it takes for
+// leader election to fail over a new replica.
+const defaultRequeueDuration = 15 * time.Second
+
+// WithLeadingManager returns a decorating reconcile.Reconciler that discards reconciliation requests
+// for the controllers that are started with the controller.Options.NeedLeaderElection
+// option set to false in non-leading replicas.
+//
+// Starting controllers in non-leading replicas is needed for these that update the data
+// served by the visibility extension API server.
+//
+// This enables to:
+//   - Keep the scheduling decisions under the responsibility of the leading replica alone,
+//     to prevent any concurrency issues.
+//   - Consume requests from the watch event queues, to prevent them from growing indefinitely
+//     in the non-leading replicas.
+//   - Transition to actually reconciling requests in the replica that may acquire
+//     the leader election lease, in case the previously leading replica failed to renew it.
+func WithLeadingManager(mgr ctrl.Manager, reconciler reconcile.Reconciler, obj client.Object, cfg *config.Configuration) reconcile.Reconciler {
+	// Do not decorate the reconciler if leader election is disabled
+	if cfg.LeaderElection == nil || !ptr.Deref(cfg.LeaderElection.LeaderElect, false) {
+		return reconciler
+	}
+
+	// Default to the recommended lease duration, that's used for core components
+	requeueDuration := defaultRequeueDuration
+	// Otherwise used the configured lease duration for the manager
+	zero := metav1.Duration{}
+	if duration := cfg.LeaderElection.LeaseDuration; duration != zero {
+		requeueDuration = duration.Duration
+	}
+
+	return &leaderAwareReconciler{
+		elected:         mgr.Elected(),
+		client:          mgr.GetClient(),
+		delegate:        reconciler,
+		object:          obj,
+		requeueDuration: requeueDuration,
+	}
+}
+
+type leaderAwareReconciler struct {
+	elected         <-chan struct{}
+	client          client.Client
+	delegate        reconcile.Reconciler
+	object          client.Object
+	requeueDuration time.Duration
+}
+
+var _ reconcile.Reconciler = (*leaderAwareReconciler)(nil)
+
+func (r *leaderAwareReconciler) Reconcile(ctx context.Context, request reconcile.Request) (reconcile.Result, error) {
```

Comment:
I think this isn't very complex, but non-trivial either, and is not covered with e2e tests. Should we have some unit tests at least for the function (we could use fake clock and fake client)? WDYT @alculquicondor @astefanutti ?

Reply:
I think the most valuable thing would be an E2E test that terminates one of the kueue instances and checks that scheduling continues happening.

Reply:
I'd also vote for having an e2e test, that'd scale to two replicas, assert the webhooks and the visibility API are correctly served in both replicas, then terminate the leading replica and scale down to one replica, check leadership has failover to the other replica, and assert scheduling and serving work as expected.

Reply:
just a nit question, why "and scale down to one replica" is relevant in the scenario? I guess it would be closer to real-life scenario if we continue to have 2 replicas,

Reply:
I'm actually happy with the simplicity of this implementation.

I think we can leave the E2E test for a follow up.

/hold cancel
/lgtm
/approve

Reply:
@astefanutti you may want to open an issue for the follow up test, so that we don't forget about it? Unless you are about to open the PR, then probably it does not matter.

Reply:
@mimowo sure, I've just created #1655 to track it.